app/models/daos/OpenIDInfoDAO.scala

```diff
@@ -0,0 +1,117 @@
+  } yield dbOpenIDInfo
+
+  protected def addAction(loginInfo: LoginInfo, authInfo: OpenIDInfo) =
+    loginInfoQuery(loginInfo).result.head.flatMap { dbLoginInfo =>
+      DBIO.seq(
+        slickOpenIDInfos += DBOpenIDInfo(authInfo.id, dbLoginInfo.id.get),
+        slickOpenIDAttributes ++= authInfo.attributes.map {
+          case (key, value) => DBOpenIDAttribute(authInfo.id, key, value)
+        })
+    }.transactionally
+
+  protected def updateAction(loginInfo: LoginInfo, authInfo: OpenIDInfo) =
+    openIDInfoQuery(loginInfo).result.head.flatMap { dbOpenIDInfo =>
+      DBIO.seq(
+        slickOpenIDInfos filter(_.id === dbOpenIDInfo.id) update dbOpenIDInfo.copy(id = authInfo.id),
+        slickOpenIDAttributes.filter(_.id === dbOpenIDInfo.id).delete,
+        slickOpenIDAttributes ++= authInfo.attributes.map {
+          case (key, value) => DBOpenIDAttribute(authInfo.id, key, value)
+        })
+    }.transactionally
+
+  /**
+   * Finds the auth info which is linked with the specified login info.
+   *
+   * @param loginInfo The linked login info.
+   * @return The retrieved auth info or None if no auth info could be retrieved for the given login info.
+   */
+  def find(loginInfo: LoginInfo): Future[Option[OpenIDInfo]] = {
+    val query = openIDInfoQuery(loginInfo).joinLeft(slickOpenIDAttributes).on(_.id === _.id)
+    val result = db.run(query.result)
+    result.map { openIDInfos =>
+      if (openIDInfos.isEmpty) None
+      else {
+        val attrs = openIDInfos.collect { case (id, Some(attr)) => (attr.key, attr.value) }.toMap
+        Some(OpenIDInfo(openIDInfos.head._1.id, attrs))
+      }
+    }
+  }
+
+  /**
+   * Adds new auth info for the given login info.
+   *
+   * @param loginInfo The login info for which the auth info should be added.
+   * @param authInfo The auth info to add.
+   * @return The added auth info.
+   */
+  def add(loginInfo: LoginInfo, authInfo: OpenIDInfo): Future[OpenIDInfo] =
+    db.run(addAction(loginInfo, authInfo)).map(_ => authInfo)
+
+  /**
+   * Updates the auth info for the given login info.
+   *
+   * @param loginInfo The login info for which the auth info should be updated.
+   * @param authInfo The auth info to update.
+   * @return The updated auth info.
+   */
+  def update(loginInfo: LoginInfo, authInfo: OpenIDInfo): Future[OpenIDInfo] =
+    db.run(updateAction(loginInfo, authInfo)).map(_ => authInfo)
+
+  /**
+   * Saves the auth info for the given login info.
+   *
+   * This method either adds the auth info if it doesn't exists or it updates the auth info
+   * if it already exists.
+   *
+   * @param loginInfo The login info for which the auth info should be saved.
+   * @param authInfo The auth info to save.
+   * @return The saved auth info.
+   */
+  def save(loginInfo: LoginInfo, authInfo: OpenIDInfo): Future[OpenIDInfo] = {
+    val query = loginInfoQuery(loginInfo).joinLeft(slickOpenIDInfos).on(_.id === _.loginInfoId)
+    val action = query.result.head.flatMap {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Usage of head on collections.](https://www.codacy.com/app/HiP-App/HiPCMS/file/2281812354/issues/source?bid=3144114&fileBranchId=3144125#l93)